refactor(traverse): revert changes to `walk.rs`

overlookmotel opened the pull request:
Revert the changes to `oxc_traverse` made in #5577.

Those changes may well have been good, but...

The soundness of `Traverse` is a finely balanced thing, and the pointer gymnastics are subtle, so I think it's best to be extremely conservative about changes. I last ran Miri on it with the original formulation (`as *mut _` etc) and it passed. Currently it's not possible to usefully run Miri on the transformer because it contains known unsound code (#3483). So until we're able to check soundness with Miri, I think it's best to avoid changes, as it'd be easy to trigger UB unintentionally.

Boshen commented:
Wait ... look at this!

https://doc.rust-lang.org/1.81.0/src/core/ptr/mod.rs.html#792-804

```
/// Convert a mutable reference to a raw pointer.
///
/// This is equivalent to `r as *mut T`, but is a bit safer since it will never silently change
/// type or mutability, in particular if the code is refactored.
#[inline(always)]
#[must_use]
#[stable(feature = "ptr_from_ref", since = "1.76.0")]
#[rustc_const_stable(feature = "ptr_from_ref", since = "1.76.0")]
#[rustc_allow_const_fn_unstable(const_mut_refs)]
#[rustc_never_returns_null_ptr]
pub const fn from_mut<T: ?Sized>(r: &mut T) -> *mut T {
    r
}
```

> This is equivalent to r as *mut T, but is a bit safer since it will never silently change type or mutability, in particular if the code is refactored.

Are there misinformation or missing context?

overlookmotel commented:
I totally agree `Traverse` should use `ptr::from_mut`, `ptr::cast` and `ptr::addr_of` etc in place of the existing `as` casting. As you say, according to the docs, it'd be a definite improvement.

It's just that I would prefer to be ultra-conservative about any changes to `Traverse` until we can run Miri on the transformer. If (as I fear) Miri *does* find problems, it will be easier to figure them out and fix them if little has changed since I last ran Miri on it.

Background is that when I originally wrote `Traverse`, it took me about a week of fiddling with tiny changes before I could get Miri to give it a pass. And when I was reading about the "rules" of what's UB and what's not, it was surprising how often you get to a point where even the experts say "I can't say if that's sound or not. It depends on what we decide the memory model is. Don't know.". There are subtleties beyond what the docs say, and I certainly don't understand it all.

So, after that experience, I'm pretty wary, and consider making *any* changes without Miri to check it kind of "flying blind".

On the positive side, there's only a handful of `AstBuilder::copy` calls left in the transformer, and I'd hope to have removed them all within next few weeks as I go though doing a "second pass" over transformer. So my preference would be to delay making these changes to `Traverse` until that's complete and we can get Miri running again.

Does that sound OK to you?

overlookmotel committed:
refactor(traverse): revert changes to `walk.rs` (#5652)

Revert the changes to `oxc_traverse` made in #5577.

Those changes may well have been good, but...

The soundness of `Traverse` is a finely balanced thing, and the pointer gymnastics are subtle, so I think it's best to be extremely conservative about changes. I last ran Miri on it with the original formulation (`as *mut _` etc) and it passed. Currently it's not possible to usefully run Miri on the transformer because it contains known unsound code (#3483). So until we're able to check soundness with Miri, I think it's best to avoid changes, as it'd be easy to trigger UB unintentionally.
